src/vmm/src/arch/aarch64/regs.rs

```diff
@@ -157,265 +110,685 @@ arm64_sys_reg!(ID_AA64MMFR2_EL1, 3, 0, 0, 7, 2);
+            index: 0,
+            offset: 0,
+            vec: self,
+        }
+    }
+
+    /// Returns an iterator over stored registers that allows register modifications.
+    fn iter_mut(&mut self) -> impl Iterator<Item = Aarch64RegisterRefMut> {
+        Aarch64RegisterVecIteratorMut {
+            index: 0,
+            offset: 0,
+            vec: self,
+        }
+    }
+}
+
+/// Wrapper type around `Aarch64RegisterVec`.
+/// Needed to ensure correctness of inner state after
+/// deserialization.
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct Aarch64RegisterVec {
+    inner: Aarch64RegisterVecInner,
+}
+
+impl Aarch64RegisterVec {
+    /// Returns the number of elements in the vector.
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns true if the vector contains no elements.
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    /// Appends a register to the vector, copping register data.
+    pub fn push(&mut self, reg: Aarch64RegisterRef<'_>) {
+        self.inner.push(reg);
+    }
+
+    /// Returns an iterator over stored registers.
+    pub fn iter(&self) -> impl Iterator<Item = Aarch64RegisterRef> {
+        self.inner.iter()
+    }
+
+    /// Returns an iterator over stored registers that allows register modifications.
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = Aarch64RegisterRefMut> {
+        self.inner.iter_mut()
+    }
 }
 
-/// Read the MPIDR - Multiprocessor Affinity Register.
-pub fn get_mpidr(vcpufd: &VcpuFd) -> Result<u64, Error> {
-    match vcpufd.get_one_reg(MPIDR_EL1) {
-        Err(err) => Err(Error::GetOneReg(MPIDR_EL1, err)),
-        // MPIDR register is 64 bit wide on aarch64, this expect cannot fail
-        // on supported architectures
-        Ok(val) => Ok(val.try_into().expect("MPIDR register to be 64 bit")),
+impl Versionize for Aarch64RegisterVec {
+    fn serialize<W: std::io::Write>(
+        &self,
+        writer: &mut W,
+        version_map: &VersionMap,
+        target_version: u16,
+    ) -> VersionizeResult<()> {
+        self.inner.serialize(writer, version_map, target_version)
+    }
+
+    fn deserialize<R: std::io::Read>(
+        reader: &mut R,
+        version_map: &VersionMap,
+        source_version: u16,
+    ) -> VersionizeResult<Self>
+    where
+        Self: Sized,
+    {
+        let inner = Aarch64RegisterVecInner::deserialize(reader, version_map, source_version)?;
+        let total_size: u64 = inner.ids.iter().map(|id| reg_size(*id)).sum();
+        if total_size as usize != inner.data.len() {
+            Err(VersionizeError::Deserialize(
+                "Failed to deserialize aarch64 registers. Sum of registers sizes is not equal to \
+                 registers data length"
+                    .to_string(),
+            ))
+        } else {
+            Ok(Self { inner })
+        }
+    }
+
+    fn version() -> u16 {
+        Aarch64RegisterVecInner::version()
```

Comment:
Mh, I think using a versionize `de_fn` might be simpler here? The documentation says they can be used for "semantic validation" [[1]], so I think we can just do the check there instead of needing another wrapper.

[1]: https://github.com/firecracker-microvm/versionize_derive/blob/f0a7194791c5cbeeb55aecfa6e60cca6f2c4168a/src/lib.rs#L169

Reply:
`de_fn` is `the method is called if the field is skipped from deserialization because it does not exist in the source version of the serialized structure`. So if will not work for snapshots with latest version.

Reply:
ugh, versionize strikes again. nvm then